torch/_inductor/index_propagation.py

```diff
@@ -267,15 +287,37 @@ def inner(*args: Any, **kwargs: Any) -> IndexPropResult:
 
         return inner
 
+    def statically_true(self, e):
+        evaluated = self.shape_env._maybe_evaluate_static(
+            e,
+            axioms=self.axioms,
+            var_to_range=self.var_to_range,
+        )
+        return bool(evaluated)
+
     def indirect_indexing(
         self, index: Union[Any, IndexPropVar], size: Any, check: bool = True
     ) -> Any:
-        # nb. We do index + Where(...) rather than Where(idx >= 0, idx, idx + sz) because we don't have CSE
-        #     for SymPy expressions, so we don't want to repeat idx too much
-
-        # indirect_indexing returns a sympy value, so no need to wrap in IndexPropVar here
         if isinstance(index, IndexPropVar) and index.is_symbolic:
-            # If we are turning a indirect indexing into direct, we need to wrap it.
-            index = index.value.expr
-            return index + Where(index >= 0, 0, size)
+            expr = sympy.sympify(index.value.expr)
+
+            def wrap_expr(expr):
+                # Positive, negative, mixed
+                if self.statically_true(0 <= expr):
+                    return expr
+                elif self.statically_true(expr < 0):
+                    return expr + size
+                else:
+                    return Where(expr < 0, expr + size, expr)
+
+            # Trivial case
+            if not generate_assert(check):
+                return wrap_expr(expr)
+
+            # It is often 's easier for us to prove that 0 <= expr, than to prove -size <= expr
+            # when dynamic shapes are on
+            if (
+                self.statically_true(0 <= expr) or self.statically_true(-size <= expr)
+            ) and self.statically_true(expr < size):
+                return wrap_expr(expr)
```

Comment:
We shouldn't just fall back to indirect indexing. We should either separate out the assertion code from indirect indexing, or just emit a warning that bounds will not be checked at runtime or something.

Reply:
I don't think I understand what you mean

Reply:
We shouldn't re-introduce the indirect indexing here, we need to either:
1. Assert the variables are in-bounds without indirect indexing, OR
2. Continue to not assert but at least warn that we are diverging from eager

Reply:
We are not re-introducing any direct indexing here? This path (at least the one you have cited) wraps the expression and converts it into a sympy expression.